pain-points.md

```diff
@@ -0,0 +1,147 @@
+repo; manually run their scraper; supply private environment
+variables; set the scraper to auto-run; add a webhook for callbacks
+post-run.
+
+Frontend also allows the user to see the console output of a scraper -
+live, if it's currently running; or the stored logs from the previous
+run. The live log streaming is mediated by
+[Faye](https://github.com/faye/faye-websocket-ruby)
+
+## mitmproxy
+
+All HTTP/S traffic from the scrapers is pushed through a single
+mitmproxy instance, running in Docker. This tracks which websites the
+scraper scrapers (eg - [the Bega Valley
+scraper](https://morph.io/planningalerts-scrapers/Bega_Valley_Applications_on_Exhibition)
+`Scrapes datracker.begavalley.nsw.gov.au`)
+
+
+# Current pain points and scaling challenges
+
+## Single machine
+
+Currently, all of Morph runs on a single Linode 32Gb VM. Obviously
+this is a single point of failure; but it also means that the number
+of concurrent scrapers is limited by the available RAM. Each scraper
+container is capped at 512Mb and we have a limit of 30 concurrent
+jobs.
+
+## Memory contention + exponential backoff = every-growing retry queue
+
+Until recently, this VM only had 224Gb RAM. It would relatively often
+(once a month or so) get into a state where Sidekiq was unable to
+schedule new tasks, because all 30 slots were full. The tasks that
+couldn't be executed would be put on a retry queue. However, as time
+went on, more tasks would be attempted, would fail, and would be put
+on the retry queue. Each time the same task failed, Sidekiq would use
+an exponential-backoff to increase the time before it was attempted
+again. This would mean that if no jobs could be scheduled for a small
+amount of time, a large amount of jobs would be in the retry
+queue. Later on, they'd all eb retried at once and the slots would
+fill; then the retry queue would get larger. This would continue until
+the retry queue contained thousands of jobs, all of which had
+hours-long retry times, and the queue would continue to get longer.
+
+If caught in time, we could manually keep retrying jobs every few
+minutes in order to pump them through faster than the exponential
+backoff was allowing - with an eye on the capacity graphs, to make
+sure we weren't retrying them too fast.
+
+Prior to the upgrade, an early sign of impending problems was
+buffers/cache dropping to 0Mb; now that we have 32Gb we consistently
+have around 8-12Gb of buffers/cache in use. We haven't seen a repeat
+of the spiral of death since the upgrade to 32Gb.
+
+## Logging overhead
+
+Currently we have to schedule and run a parallel job for every docker
+container to scrape its console output and feed it to the DB and
+Faye. These jobs time out, so we have to have another job which looks
+for running containers without a log-scraper and start a new
+log-scraping job. This could probably be avoided if we could do
+something like feed the logs into syslog or some other central logging
+system.
+
```

Comment:
## mitmproxy troubleshooting

Because mitmproxy intercepts HTTPS traffic, we've got a self-signed CA cert on it which needs to be trusted by the scrapers. Most of the customisation on our buildstep image is trying to get this cert in all the places and trusted by all the things.

However, the dev vagrant image doesn't (unless my memory is playing tricks on me) set this up at all; which makes it hard to reproduce problems. Having a dev setup that did set up SSL by default - in as close as possible to the prod setup - would make it easier to troubleshoot problems.

Stepping back a little though: we really only have mitmproxy in place so that we can record the hosts the scraper is scraping from. If we had a different way to record those hosts, we might not need mitmproxy at all.

## Scraper platform updates

At present all scrapers use `openaustralia/buildstep:latest`. This image is based on `cedar:14`, which is old, doesn't work at all with PHP, and about to be EOL. We need to update to `heroku:16`, but this is a breaking change: different set of system packages installed, different versions of languages avalable and so on. [#1207](https://github.com/openaustralia/morph/pull/1207) is the start of a plan to allow scrapers to pick a platform so that we can start with a soft migration to heroku:16, and later make other platforms available.